patches/bess/0012-ExactMatch-Bulk-Support.patch

```diff
@@ -1,15 +1,53 @@
-From 460ac23d8143f8475a7819aa8ccbbf930f2940c8 Mon Sep 17 00:00:00 2001
-From: Amar Srivastava <amarsri28@gmail.com>
-Date: Wed, 23 Dec 2020 02:33:49 -0800
-Subject: [PATCH] dpdk-hash linking and pipeline improvements
+From 54cd30fdc2cac86900d564e90118b7fbef7bef20 Mon Sep 17 00:00:00 2001
+From: Amar Sri <amarsri28@gmail.com>
+Date: Thu, 4 Mar 2021 06:33:08 -0800
+Subject: [PATCH] Exact match bulk support
 
 ---
+ core/modules/exact_match.cc    |  25 ++--
  core/modules/wildcard_match.cc | 202 ++++++++++++++++++++++++++-------
  core/modules/wildcard_match.h  |  37 ++++--
  core/utils/cuckoo_map.h        | 146 +++++++++++++++++++++++-
- core/utils/exact_match_table.h |  97 +++++++++++-----
- 4 files changed, 403 insertions(+), 79 deletions(-)
+ core/utils/exact_match_table.h |  92 ++++++++++-----
+ 5 files changed, 410 insertions(+), 92 deletions(-)
 
+diff --git a/core/modules/exact_match.cc b/core/modules/exact_match.cc
+index 1ca0fd4e..0d978518 100644
+--- a/core/modules/exact_match.cc
++++ b/core/modules/exact_match.cc
+@@ -326,17 +326,22 @@ void ExactMatch::ProcessBatch(Context *ctx, bess::PacketBatch *batch) {
+ 
+   int cnt = batch->cnt();
+   Value default_value(default_gate);
+-
+-  for (int i = 0; i < cnt; i++) {
+-    bess::Packet *pkt = batch->pkts()[i];
+-    ValueTuple res;
+-    res = table_.Find(keys[i], default_value);
+-    if (res.gate != default_gate) {
+-      /* setting respecive values */
+-      setValues(pkt, res.action);
++   ValueTuple *res[32]; 
++  for (int i = 0; i < cnt; ) {
++    int num= (cnt-i >= 32) ?32:cnt-i;
++    uint64_t hit_mask = table_.Find(&keys[i],&res[0],num);
++
++   for(int j=i;j<num+i;j++)
++   {
++		if ((hit_mask & (1ULL << (j-i)))  == 0) 
++      EmitPacket(ctx, batch->pkts()[j],default_gate);
++    else
++    {
++	    setValues(batch->pkts()[j], res[j-i]->action);
++      EmitPacket(ctx, batch->pkts()[j], res[j-i]->gate);
+     }
+-    gate_idx_t g = res.gate;
+-    EmitPacket(ctx, pkt, g);
++   }
++   i = i+num;
+   }
```

Comment:
Any reason to expose 32 here? Can we simplify it to be like below and move the detail all the way down to cuckooMap lookup_bulk_data API? Since at that level you can do sub-batch iteration using [`RTE_HASH_LOOKUP_BULK_MAX`](https://doc.dpdk.org/api/rte__hash_8h.html#a40d9cbb1b31cf0a18427128378bc94ec) constraint

```cpp
// Ensure max burst size fits in the hitmask
static_assert(bess::PacketBatch::kMaxBurst <= 64);

ValueTuple *res[bess::PacketBatch::kMaxBurst];
uint64_t hit_mask = table_.Find(&keys, &res, cnt);

// Iterate results and set values and gate
for(int i = 0; i < cnt; i++)
{
    if(hit_mask & (1ULL << i)) {
        setValues(batch->pkts()[i], res[i]->action);
        EmitPacket(ctx, batch->pkts()[i], res[i]->gate);
    }
    else {
        EmitPacket(ctx, batch->pkts()[i],default_gate);
    }
}
```

Reply:
a more general question NOT pertaining to this PR, The `ValueTuple` being pointed could be in the process of updating by the writer correct? How do we handle that scenario?

Reply:
slow path forward separate event for update , it can be taken care.

Reply:
I did not understand. We have a lock for the key but no lock while accessing the value.

Reply:
yes there is no lock for accessing the value so its cleaner to handle 'update' separately .